src/main/java/org/opensearch/knn/training/TrainingJobRunner.java

```diff
@@ -176,4 +193,122 @@ private void serializeModel(TrainingJob trainingJob, ActionListener<IndexRespons
     public int getJobCount() {
         return jobCount.get();
     }
+
+    /**
+     * This method is called whenever the cluster state changes. It is used to update models that are still training when a node leaves or the cluster crashes.
+     * @param event the event that changed the cluster change
+     */
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.localNodeClusterManager()) {
+            if (event.isNewCluster()) {
+                threadPool.schedule(() -> {
+                    try {
+                        updateModelsNewCluster();
+                    } catch (IOException | InterruptedException | ExecutionException e) {
+                        throw new RuntimeException(e);
+                    }
+                }, TimeValue.timeValueSeconds(1), ThreadPool.Names.GENERIC);
+            } else if (event.nodesRemoved()) {
+                List<DiscoveryNode> removedNodes = event.nodesDelta().removedNodes();
+                threadPool.schedule(() -> {
+                    try {
+                        updateModelsNodesRemoved(removedNodes);
+                    } catch (IOException | InterruptedException | ExecutionException e) {
+                        throw new RuntimeException(e);
+                    }
+                }, TimeValue.timeValueSeconds(0), ThreadPool.Names.GENERIC);
+            }
+        }
+    }
+
+    protected void updateModelsNewCluster() throws IOException, InterruptedException, ExecutionException {
+        if (modelDao.isCreated()) {
+            List<String> modelIds = searchModelIds();
+            for (String modelId : modelIds) {
+                Model model = modelDao.get(modelId);
+                ModelMetadata modelMetadata = model.getModelMetadata();
+                if (modelMetadata.getState().equals(ModelState.TRAINING)) {
+                    modelMetadata.setState(ModelState.FAILED);
+                    modelMetadata.setError("Training failed to complete due to node drop");
+                    modelDao.update(model, new ActionListener<IndexResponse>() {
+                        @Override
+                        public void onResponse(IndexResponse indexResponse) {
+                            logger.info(
+                                "Model "
+                                    + indexResponse.getId()
+                                    + " updated from "
+                                    + ModelState.TRAINING
+                                    + " to "
+                                    + ModelState.FAILED
+                                    + " due to node drop"
+                            );
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            logger.info("Failed to update model after node drop", e);
```

Comment:
For log messages, could we add details as to which model failed for a particular action?

Reply:
Should be logged as error level.